### ISSUE TITLE
laggy camera

### ISSUE THREAD

**Issue by YMZnew** [EdwardLu2018/wasm-ar#1]
thank you for sharing your experience .
I was testing your code , as the time I click on screen , camera came so laggy with 2FPS . 
while tracking is well , i think there is sth wrong in detection when click screen or on repeat the detection after lost tracker .
would you please help me solve this issue ?

tested on : 
samsung note 10 plus
surface pro 3 camera
samsung A50

**Comment by EdwardLu2018** [comment]
When you press the screen, it enables detection and tracking, so it will constantly be detecting until it finds a valid detection, then it will start tracking. Detection (which consists of feature point extraction and matching) is slow, even in WASM, so it will definitely lag your camera trying to find the image. I believe what you are seeing is not a bug, but just a desired behavior. You can turn off detection and tracking manually by pressing the screen again. When I get the time, I can implement a more intelligent way to determine when the image is lost and turn off detection automatically.

**Comment by YMZnew** [comment]
thank you very much for your answer . I've researched for so many methods for detection and their complexity and I think using NFT ( natural featuee tracking ) markers can improve the performance . 
what do you think about NFT markers?

**Comment by kalwalt** [comment]
Hi @EdwardLu2018 and @YMZnew i think that fps can be improved a bit using a worker, as we ( me and @ThorstenBux) did in [jsartoolkit5](https://github.com/artoolkitx/jsartoolkit5 ) for NFT marker detection and tracking. See [threejs_wasm_worker.js](https://github.com/artoolkitx/jsartoolkit5/blob/master/examples/nft_improved_worker/threejs_wasm_worker.js) and the [worker](https://github.com/artoolkitx/jsartoolkit5/blob/master/js/artoolkit.worker.js); i was thinking to  try this approach also with your code. If you are interested we have founded a new github org for WebAR https://github.com/webarkit

**Comment by ThorstenBux** [comment]
Nice one, also if you look into artoolkitX.js and the wasm binding there I highly optimised the video hand over . Maybe that is a lag too.

Get Outlook for iOS<https://aka.ms/o0ukef>
________________________________

**Comment by kalwalt** [comment]
Yes @ThorstenBux you are right! Here you can find [ArtoolkitX.js](https://github.com/webarkit/artoolkitX.js)

**Comment by EdwardLu2018** [comment]
Woah! Thanks for all the great comments everyone! Did not expect this project to get this much attention, ha!

@YMZnew I'll be sure to check out NFT tracking! I hear AR.js has an implementation running on the web. Not sure if its running in WASM, though.

@kalwalt I'll definitely try using a web worker when I get the chance. I've used workers in another, similar project I did but never got around to adapting it for this one. I believe I used Comlink in that other project, so I'll probably do the same here eventually.
